File: 127. Word Ladder/ladderLength.py
from typing import List

class Solution:
  def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
    if endWord not in wordList:
      return 0

    neighbors = {}
    wordList.append(beginWord)
    for word in wordList:
      for i in range(len(word)):
        key = word[:i] + '*' + word[i + 1:]
        if key not in neighbors:
          neighbors[key] = []
        neighbors[key].append(word)

    queue = [(beginWord, 1)]
    visited = set([beginWord])

    while queue:
      word, level = queue.pop(0)
      for i in range(len(word)):
        key = word[:i] + '*' + word[i + 1:]
        for neighbor in neighbors[key]:
          if neighbor == endWord:
            return level + 1
          if neighbor not in visited:
            visited.add(neighbor)
            queue.append((neighbor, level + 1))

    return 0
  

beginWord = "hit"
endWord = "cog"
wordList = ["hot","dot","dog","lot","log","cog"]
solution = Solution()
print('ladderLength: ', solution.ladderLength(beginWord, endWord, wordList)) # 5

        

File: 127. Word Ladder/ladderLength.js
/**
 * @param {string} beginWord
 * @param {string} endWord
 * @param {string[]} wordList
 * @return {number}
 */
const ladderLength = (beginWord, endWord, wordList) => {
  const neibors = {};
  wordList.push(beginWord);

  for (let i = 0; i < wordList.length; i++) {
    const word = wordList[i];
    for (let j = 0; j < word.length; j++) {
      const key = word.slice(0, j) + '*' + word.slice(j + 1);
      if (!neibors[key]) {
        neibors[key] = [];
      }
      neibors[key].push(word);
    }
  }

  const queue = [[beginWord, 1]];
  const visited = new Set();

  while (queue.length) {
    const [word, level] = queue.pop();
    for (let i = 0; i < word.length; i++) {
      const key = word.slice(0, i) + '*' + word.slice(i + 1);
      const wordNeibors = neibors[key];
      for (let j = 0; j < wordNeibors.length; j++) {
        if (wordNeibors[j] === endWord) {
          return level + 1;
        }
        if (!visited.has(wordNeibors[j])) {
          visited.add(wordNeibors[j]);
          queue.push([wordNeibors[j], level + 1]);
        }
      }
    }
  }

  return 0;
};

const beginWord = 'hit';
const endWord = 'cog';
const wordList = ['hot', 'dot', 'dog', 'lot', 'log', 'cog'];
console.log("ladderLength: ", ladderLength(beginWord, endWord, wordList)); // 5